File: apps/dashboard/types/campaign.ts
export type CampaignStatus = "draft" | "active" | "paused" | "completed";

export type CreativeType = "html5" | "image" | "video";

export interface CampaignTargeting {
  geo: string[];
  platforms: string[];
  age_range?: { min: number; max: number };
  interests: string[];
}

export interface Campaign {
  id: string;
  org_id: string;
  name: string;
  status: CampaignStatus;
  targeting: CampaignTargeting;
  budget_cents: number | null;
  currency: string;
  start_date: string | null;
  end_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface Creative {
  id: string;
  org_id: string;
  campaign_id: string;
  name: string;
  type: CreativeType;
  file_url: string;
  file_size_bytes: number | null;
  preview_url: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateCampaignRequest {
  name: string;
  targeting?: Partial<CampaignTargeting>;
  budget_cents?: number | null;
  currency?: string;
  start_date?: string | null;
  end_date?: string | null;
}

export interface UpdateCampaignRequest {
  name?: string;
  targeting?: Partial<CampaignTargeting>;
  budget_cents?: number | null;
  currency?: string;
  start_date?: string | null;
  end_date?: string | null;
}

export interface UpdateCampaignStatusRequest {
  status: CampaignStatus;
}

export interface CreateCreativeRequest {
  name: string;
  type: CreativeType;
  file_url: string;
  file_size_bytes?: number | null;
  preview_url?: string | null;
}

export interface CampaignListResponse {
  items: Campaign[];
  total: number;
  limit: number;
  offset: number;
}

export interface CreativeListResponse {
  items: Creative[];
  total: number;
}

export const VALID_TRANSITIONS: Record<CampaignStatus, CampaignStatus[]> = {
  draft: ["active"],
  active: ["paused", "completed"],
  paused: ["active", "completed"],
  completed: [],
};

File: services/api-dashboard/internal/model/campaign.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type CampaignStatus string

const (
	StatusDraft     CampaignStatus = "draft"
	StatusActive    CampaignStatus = "active"
	StatusPaused    CampaignStatus = "paused"
	StatusCompleted CampaignStatus = "completed"
)

// ValidTransitions defines the allowed next states for each campaign status.
// A terminal state (completed) has no allowed transitions.
var ValidTransitions = map[CampaignStatus][]CampaignStatus{
	StatusDraft:     {StatusActive},
	StatusActive:    {StatusPaused, StatusCompleted},
	StatusPaused:    {StatusActive, StatusCompleted},
	StatusCompleted: {},
}

type AgeRange struct {
	Min int `json:"min"`
	Max int `json:"max"`
}

type CampaignTargeting struct {
	Geo       []string  `json:"geo"`
	Platforms []string  `json:"platforms"`
	AgeRange  *AgeRange `json:"age_range,omitempty"`
	Interests []string  `json:"interests"`
}

// Campaign represents a marketing campaign.
  type Campaign struct {
	ID          uuid.UUID         `json:"id"`
	OrgID       uuid.UUID         `json:"org_id"`
	Name        string            `json:"name"`
	Status      CampaignStatus    `json:"status"`
	Targeting   CampaignTargeting `json:"targeting"`
	BudgetCents *int64            `json:"budget_cents"`
	Currency    string            `json:"currency"`
	StartDate   *time.Time        `json:"start_date"`
	EndDate     *time.Time        `json:"end_date"`
	CreatedAt   time.Time         `json:"created_at"`
	UpdatedAt   time.Time         `json:"updated_at"`
}
